### DISCUSSION TITLE
2021-09-08 (automatic release)

### DISCUSSION THREAD

**Discussion by pipcet** [pipcet/pearl#22] [General]
This release continues work on the boot and installation process. The major addition is code to make use of, and set, an nvram variable called `pearl` which specifies commands to be executed during the boot process.

Details and instructions in the discussion!

**Comment by pipcet** [comment]
Instructions

To install pearl, boot into Recovery Mode, select Options, then Terminal from (which?) menu, then enter:

```
curl -L https://github.com/pipcet/pearl/releases/latest/download/pearl.pl|perl
```

You will be prompted for boot parameters which are stored in the nvram variable `pearl`. Useful ones include specifying an autoboot delay (default is 60 seconds) with `delay 15` (for 15 seconds), a root file system with `root /dev/mapper/vg-vol0` (to boot an uninventively named volume `vol0` in volume group `vg`). On the mini, `no4k` may be useful to prevent the automatic switch to 4k resolution, and `gadget-mode 0` or `gadget-mode 1` can be used to force one of the USB ports into gadget mode so you can control pearl remotely via a USB cable. Multiple options can be combined by separating with `;`.

After that, you will be asked to confirm installation by entering the username and password of an administration user. Please be sure at this point that you're aware that I'm unable to make warranties of any kind. Also remember that you need to use `bputil -n` in the Recovery Mode terminal to return to macOS booting.

The machine should reboot automatically, boot to the pearl screen, and then either attempt to boot from the device you specified with `root /dev/mapper/vg-vol0` above or launch a ramdisk containing a base Debian system. In the latter case, `root` will be logged in automatically and be asked to configure the network connection (if manual IP configuration is required, terminate the connection dialog and enter `nmtui` instead). After that, you should have a working system, but please remember that it is based on a ramdisk and thus changes you make are not persistent. It is also quite easy to run out of RAM, so you might want to consider installing on a USB device next.

X-based desktop environments should work. (Other desktop environments may or may not work.)

**Reply by Johnnie390** [reply]
Hello all/Pip,

as per your instructions above. Bputil -nkcas and csrutil disable commands had already been issued before the sequence of commands mentioned below - 

-bash-3.2# curl -L https://github.com/pipcet/pearl/releases/latest/download/pearl.pl|perl
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   135  100   135    0     0    490      0 --:--:-- --:--:-- --:--:--   490
100   616  100   616    0     0   1371      0 --:--:-- --:--:-- --:--:--  1371
100  240M  100  240M    0     0  2843k      0  0:01:26  0:01:26 --:--:-- 2237k
tar: Removing leading '/' from member names
Available boot parameters (separate with ';'):
  no4k            disable automatic switch to 4k resolution
  root DEV        automatically boot from device
  gadget-mode 0   put USB port 0 (top/left) in gadget mode
  gadget-mode 1   put USB port 1 (bottom/right) in gadget mode
  delay SECS      delay n seconds before booting automatically
By setting a custom boot object, you will be putting your system into Permissive Security.

Are you sure you want to do this? (enter y or n) y
wrapping boot payload...
going to change bootpolicy to permissive...
updating local machine policy...
Username: terry
Password:
installing boot objects...
done.
**nvram: Error setting variable - 'pearl': (iokit/common) invalid argument
"Since we got here, something went wrong. Did you use bputil and csrutil?"**
-bash-3.2#

Any thoughts?

Regards,

Ry

**Reply by pipcet** [reply]
(Sorry, I'm travelling this weekend, so might not be very responsive).

Thanks for the report! Just to confirm, there was no prompt after "delay SECS" and before "By setting a custom boot object"? There should have been...

I would suggest setting the variable explicitly in the Recovery Mode shell by running
```
nvram pearl='no4k;delay 15'
```
replacing the delay specification by the right amount. Then `reboot` should work...

**Reply by Johnnie390** [reply]
Hello Pip,

I can confirm that there is _no prompt_ between "delay SECS..." and "By setting a custom boot object..." The text block above has been lifted direct from the terminal.

Regards,


Ry

**Reply by Johnnie390** [reply]
Pip,

just tried again with your latest and greatest (2021-09-11).
After boot parameter prompt - 
"Old boot parameter"
"New boot parameter"
Input was **no4k**.
After the usual startup messages/diagnostics - kernel panic.
Please see attached photo.


![WhatsApp Image 2021-09-11 at 14 08 06](https://user-images.githubusercontent.com/22815859/132947365-7383bb2f-9294-49f5-b74b-298eb21be791.jpeg)


Regards,

Ry

**Comment by Johnnie390** [comment]
Hello Pip,

I see we have been busy...
I presume the same command string - curl -L https://github.com/pipcet/pearl/releases/latest/download/pearl.pl|perl will work with today's release?

Regards,

Ry

**Reply by pipcet** [reply]
> Hello Pip,

Hello!

> I see we have been busy...

Indeed. Busy enough for me to start a new discussion at #23.

> I presume the same command string - curl -L [https://github.com/pipcet/pearl/releases/latest/download/pearl.pl|perl](https://github.com/pipcet/pearl/releases/latest/download/pearl.pl%7Cperl) will work with today's release?

It should, yes. Please report it if it doesn't.

As always, thank you for testing and the comments
Pip

**Reply by cristim** [reply]
to make it a bit more convenient I shortened the URL using bit.ly so you can use https://bit.ly/getpearl instead of the long Github URL

```
$ curl https://bit.ly/getpearl  -I 
HTTP/2 301
server: nginx
date: Wed, 15 Sep 2021 12:35:32 GMT
content-type: text/html; charset=utf-8
content-length: 152
cache-control: private, max-age=90
content-security-policy: referrer always;
location: https://github.com/pipcet/pearl/releases/latest/download/pearl.pl
referrer-policy: unsafe-url
via: 1.1 google
alt-svc: clear
```

**Reply by pipcet** [reply]
> to make it a bit more convenient I shortened the URL using bit.ly so you can use https://bit.ly/getpearl instead of the long Github URL

Thank you, that's a really good idea! I'll test

```
curl -L https://bit.ly/getpearl|perl
```

and see whether that works. If it does, I'll update the instructions.

Thanks again, and if you do get around to testing, please let me know how it goes :-)

Pip
